### ISSUE TITLE
CA1859 improperly handles explicit interface implementations

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn-analyzers#6416]
### Analyzer

**Diagnostic ID**: CA1859

### Analyzer source

https://github.com/dotnet/roslyn-analyzers/commit/6ae2683c5033dfde0b73d4e73b372f00ca6ceacd

### Describe the bug

The analyzer appears to not be factoring in whether methods from interfaces are implemented explicitly, resulting in warnings that, if fixed, will result in broken code.

### Steps To Reproduce

```C#
[Fact]
public static async Task Repro()
{
    const string Source = @"
        #nullable enable

        interface IFoo
        {
            int M();
        }

        internal class C : IFoo
        {
            int IFoo.M() => 42;
        }

        internal class Use
        {
            static int Bar()
            {
                IFoo f = new C();
                return f.M();
            }
        }
        ";

    await TestCSAsync(Source);
}
```

### Expected behavior

No warning.

### Actual behavior

```
info CA1859: Change type of variable 'f' from 'IFoo' to 'C' for improved performance
```
but if the developer tries to change the variable to type `C`, it'll fail to compile because there's no available `M()` to call.

cc: @geeknoid, @buyaa-n

**Comment by geeknoid** [comment]
Just fixed that one a few minutes ago. Will be in my next PR.
